### DISCUSSION TITLE
What automated testing is done prior to every release of the USWDS?

### DISCUSSION THREAD

**Discussion by mgifford** [uswds/uswds#5337] [Accessibility]
Are automated tools like axe, WAVE & Google Lighthouse done on components built in the design system? 

Having a design system without automated errors doesn't make it accessible, but it is useful to know that errors flagged by automated tools are not being added to the design system (even if they have been tested with assistive technology).

**Comment by thisisdano** [comment]
Hey @mgifford — every release and every push to the `uswds` repo gets a codebase scan from pa11y, which is running on aXe connected to WCAG 2.1 AA. We also run a similar scan on every push on our website (`uswds-site`).

Anything to add to this @mejiaj?

**Reply by mgifford** [reply]
Pa11y+axe is definitely a terrific practice. It only runs against static pages, so it would be important to have a demo site that shows all the components that are run. Could just run the main site but you don't want to be [running ~200 pages with every commit](https://www.google.ca/search?q=site%3Ahttps%3A%2F%2Fdesignsystem.digital.gov). Some more on [testing with GitHub Actions here.](https://accessibility.civicactions.com/posts/automated-accessibility-testing-leveraging-github-actions-and-pa11y-ci-with-axe)

What we've found is that if you select a set of URLs to crawl with Pa11y, those can get out-of-date pretty quickly. It can be really useful to see that those are tied to the patterns that are in-use. Also, so important that failures in pa11y stop commits from getting added to the repo. 

I don't know if any of the components would need cypress-axe to do the testing. We've got some [content here for cypress](https://opensource.com/article/23/2/automated-accessibility-testing). Also, some examples of automating some keyboard navigation testing there.

**Reply by mejiaj** [reply]
### General a11y testing

In USWDS we run an axe scan against components in StorybookJS via a plugin.

https://github.com/uswds/uswds/blob/218b91a2048074f534f776bdf1369cf51ae96508/package.json#L66-L67

---

### Pa11y

I really like that automation to add the pa11y notes as a comment in the PR. Testing pa11y against *every* page can be tedious and slow. It would be great to find a faster way. 

### Cypress

We already do some keyboard testing in Mocha unit tests, but definitely not the best developer experience. There can be quirks with some events, like copy/paste.

**Comment by thisisdano** [comment]
We do run it against all the pages of our site ever time, and yes it is slow! I think we're open to some improvements to our automations, and I'll check out the resources you've linked. Thanks!

**Comment by mgifford** [comment]
We've been very impressed with https://unlighthouse.dev recently. Definitely worth addign to your toolbox.

It isn't as good for accessibility as the latest axe, but does include many axe tests. 

Here's a scan of the USWDS site:

https://docs.google.com/spreadsheets/d/1IGIgC34NMJ1wEa9wYahldbphLG-vTMpvmIhV5spW6Ds/edit#gid=112532570

This is all pretty fresh though. 

Simple command line interface to dump a CSV:
`npx unlighthouse-ci --no-cache --sitemaps https://designsystem.digital.gov/sitemap.xml --site https://designsystem.digital.gov  --reporter csvExpanded`


Pretty fast for over 150 pages:
```
Unlighthouse has finished scanning https://designsystem.digital.gov: 156 routes in 357s.                      Unlighthouse 1:00:27 p.m.
✔ Generated csvExpanded report ./.unlighthouse/ci-result.csv                                                    Unlighthouse 1:00:27 p.m.
```

**Reply by mejiaj** [reply]
Would this replace pa11y or work alongside it? 

I like pa11y as a tool, but it's challenging when doing active development because it can be slow and prone to crashing (even when. messing with the settings).

**Comment by mgifford** [comment]
Along side. Pa11y generally draws from a set CSV of URLs or a sitmap URL. 

Generally we recommend to only use Pa11y with a 1-2 dozen URLs. Critical URLs and those which are being actively developed. Absoltuely do more if you can, but it shoudl be reviewed regularly to see that tests are covering what is being developed. Bugs appearing in this set list of pages shoudl stop commits getting into the project. 

Crawls can reveal different results because they are clicking on things. It is easier to run in the wee hours of the day and results can be aggregated and evalauted with every sprint. Are we doing better or worse than we were the last sprint?

**Reply by mejiaj** [reply]
I really like the idea of identifying the "core" URLs to test with pa11y and using another tool to supplement that.
